mlflow/metrics/genai/model_utils.py

```diff
@@ -146,13 +154,166 @@ def _call_openai_api(openai_uri, payload, eval_parameters, extra_headers, proxy_
 
 
 _PREDICT_ERROR_MSG = """\
-Failed to call the deployment endpoint. Please check the deployment URL\
+Failed to call the deployment endpoint. Please check the deployment URL \
 is set correctly and the input payload is valid.\n
 - Error: {e}\n
 - Deployment URI: {uri}\n
 - Input payload: {payload}"""
 
 
+def _is_supported_llm_provider(schema: str) -> bool:
+    from mlflow.gateway.provider_registry import provider_registry
+
+    return schema in provider_registry.keys()
+
+
+def _call_llm_provider_api(
+    provider_name: str,
+    model: str,
+    input_data: str,
+    eval_parameters: dict[str, Any],
+    extra_headers: dict[str, str],
+    proxy_url: Optional[str] = None,
+) -> str:
+    """
+    Invoke chat endpoint of various LLM providers.
+
+    Under the hood, this function uses the MLflow Gateway to transform the input/output data
+    for different LLM providers.
+
+    Args:
+        provider_name: The provider name, e.g., "anthropic".
+        model: The model name, e.g., "claude-3-5-sonnet"
+        input_data: The input string prompt to send to the model as a chat message.
+        eval_parameters: The additional parameters to send to the model, e.g. temperature.
+        extra_headers: The additional headers to send to the provider.
+        proxy_url: Proxy URL to be used for the judge model. If not specified, the default
+            URL for the LLM provider will be used.
+    """
+    from mlflow.gateway.config import Provider
+    from mlflow.gateway.schemas import chat
+
+    provider = _get_provider_instance(provider_name, model)
+
+    chat_request = chat.RequestPayload(
+        model=model,
+        messages=[
+            chat.RequestMessage(role="user", content=input_data),
+        ],
+        **eval_parameters,
+    )
+
+    # Filter out keys in the payload to the specified ones + "messages".
+    # Does not include "model" key here because some providers do not accept it as a
+    # part of the payload. Whether or not to include "model" key must be determined
+    # by each provider implementation.
+    filtered_keys = {"messages", *eval_parameters.keys()}
+
+    payload = {
+        k: v
+        for k, v in chat_request.model_dump().items()
+        if (v is not None) and (k in filtered_keys)
+    }
+    chat_payload = provider.adapter.chat_to_model(payload, provider.config)
+    chat_payload.update(eval_parameters)
+
+    if provider_name in [Provider.AMAZON_BEDROCK, Provider.BEDROCK]:
+        # Bedrock client doesn't support custom headers and proxy URL
+        response = provider._request(chat_payload)
```

Comment:
Could we add a warning here if extra_headers is not empty?

Reply:
Makes sense🙂